rest/client.go

```diff
@@ -0,0 +1,82 @@
+package rest
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+// Client is a wrapper around SAMS primitive REST API. Most likely, you want to use
+// the more robust service-to-service protobuf API.
+//
+// This API is needed when using SAMS to identify users, but not perform authorization
+// checks. e.g. the caller will handle its own authorization checks based on the identity
+// of the SAMS user. (The returned User.Subject, the SAMS account external ID.)
+type Client interface {
+	// GetUserDetails returns the basic user profile of the user whom the supplied SAMS access
+	// token belongs to.
+	//
+	// If the supplied token is invalid, malformed, or expired, the error will contain
+	// "unexpected status 401".
+	GetUserDetails(ctx context.Context, token string) (*User, error)
+}
+
+// NewClient constructs a new SAMS REST client, pointed to the supplied SAMS host.
+// e.g. "https://accounts.sourcegraph.com".
+func NewClient(samsHost string) Client {
+	// Canonicalize the host so we only need to check if it ends in a slash or not once.
+	samsHost = strings.ToLower(samsHost)
+	samsHost = strings.TrimSuffix(samsHost, "/")
+
+	return &client{
+		host: samsHost,
+	}
+}
+
+type client struct {
+	host string
+}
+
+func (c *client) GetUserDetails(ctx context.Context, token string) (*User, error) {
```

Comment:
nit: In general I'd say `GetUser` is concise and clear enough (to get a `User` object), unless you think it otherwise.

```suggestion
func (c *client) GetUser(ctx context.Context, token string) (*User, error) {
```

Or `GetUserInfo`, which uses a consistent term (as it is hitting the user info endpoint).

Reply:
Sounds good to me. #namingishard I wasn't sure if you would have preferred "GetOIDCUserDetails" or something more specialized. But I agree `GetUser` is the most preferable option.